rfcs/2020-11-01-remap-metrics-support.md

```diff
@@ -0,0 +1,199 @@
+# RFC 3550 - 2020-11-01 - Support metrics with the new 'remap' transform
+
+It would be useful to add the ability for the Remap language to work with metric
+data as well as with log data.
+
+## Scope
+
+The aim of the Remap language is to remain simple and yet provide the power to
+allow the user to replace most of their tranform configuration with a simple
+Remap script. Any complexity beyond a certain level is possible, but requires
+the `Lua` or `Wasm` transforms.
+
+Given this, the scope of this RFC is not to enhance the Remap language so it
+can work comprehensively with metric data. The main aim will be to allow the
+user to make simple changes to the metadata surrounding the metric.
+
+
+## Motivation
+
+Currently there are some simple transforms that allow manipulating the metadata
+of metric data - `add_tags`, `remove_tags`, `rename_tags`. These transforms
+are useful, but limited. They cannot take into account any additional data around
+the metric to determine how to process the tags.
+
+
+## Doc-level Proposal
+
+The Remap language can work with the metadata around a metric.
+
+When the event the Remap function is working with, the metric metadata can be 
+accessed through a number of set fields in a similar way to how the language 
+works with log events.
+
+The following fields are available:
+
+- `.name` - The name of the metric.
+- `.namespace` - The namespace for the metric.
+- `.timestamp` - Timestamp of the metric.
+- `.kind` - The kind of the metric - *Incremental* or *Absolute*
+- `.tags` - A map containing the tags set for the metric.
+ 
+There are strict limitations to how this data can be represented. 
+
+Setting `name` or `namespace` to anything other than a string will be an 
+error. `.timestamp` has to be a timestamp. `tags` is a map of string keys to 
+string values. No nesting is allowed. 
+
+To add a tag, or change an existing tag you can set it:
+
+```
+.tags.host = "localhost"
+```
+
+To delete a tag, use the `del` function:
+
+```
+del(".tags.host")
+```
+
+Note variables (identifiers starting with a `$`) are available and there are
+no limitations on the types you can set variables to.
+
+## Rationale
+
+Enabling the Remap language to work with metric events will allow users to
+simplify their configurations if they already use the `*_tag` transforms
+provided. It will also allow them remove the use of the `Lua` transform if they
+have been falling back to that for anything but the most complex manipulation.
+
+
+## Drawbacks
+
+In order to maintain these functions there will be a slight maintenance burden.
+
+By exposing the metric data to Remap, it may place some restrictions should we 
+want to change the internal model in the future. The more we decide to expose to
+the language, the more it may restrict us going forwards if we need to ensure
+backwards compatibility for users configurations.
```

Comment:
This is a valid point, although I think the set of attributes this RFC proposes to expose seem fairly straight-forward and not something we're likely to change.